packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts

```diff
@@ -0,0 +1,268 @@
+      endpoint: extendedEndpoint
+    });
+    expect(mockExtendedRequestMapper).toHaveReturnedWith(mappedExtendedRequest);
+
+    expect(mockExtendedHttpClient).toHaveBeenCalledTimes(1);
+    expect(mockExtendedHttpClient).toHaveBeenCalledWith(extendedEndpoint, {
+      parameters: mappedExtendedRequest
+    });
+
+    expect(mockExtendedResponseMapper).toHaveBeenCalledTimes(1);
+    expect(mockExtendedResponseMapper).toHaveBeenCalledWith(extendedRawResponse, {
+      endpoint: extendedEndpoint,
+      requestParameters: mappedExtendedRequest
+    });
+    expect(mockExtendedResponseMapper).toHaveReturnedWith(extendedResponse);
+    expect(response).toEqual(extendedResponse);
+  });
+
+  it('should use the raw request if no requestMapper is provided', async () => {
+    const { endpointAdapter, options, mockedHttpClient, mockedRequestMapper } =
+      createEndpointAdapterFactoryOptions<TestRequest, TestResponse>({
+        options: {
+          requestMapper: undefined
+        }
+      });
+
+    await endpointAdapter(request);
+
+    expect(mockedRequestMapper).not.toHaveBeenCalled();
+    expect(mockedHttpClient).toHaveBeenCalledTimes(1);
+    expect(mockedHttpClient).toHaveBeenCalledWith(options.endpoint, {
+      parameters: request
+    });
+  });
+
+  it('should return the raw response if no responseMapper is provided', async () => {
+    const rawResponse: TestResponse = { query: 'patata' };
+    const { endpointAdapter, mockedHttpClient, mockedResponseMapper } =
+      createEndpointAdapterFactoryOptions<TestRequest, TestResponse>({
+        options: {
+          responseMapper: undefined
+        },
+        rawResponse
+      });
+
+    const response = await endpointAdapter(request);
+
+    expect(mockedHttpClient).toHaveBeenCalledTimes(1);
+    expect(mockedResponseMapper).not.toHaveBeenCalled();
+    expect(response).toEqual(rawResponse);
+  });
+
+  describe('endpoint scenarios', () => {
+    // eslint-disable-next-line max-len
+    it('should use the provided EndpointAdapterOptions.endpoint directly when it is a string', async () => {
+      const { endpointAdapter, options, mockedHttpClient } = createEndpointAdapterFactoryOptions<
+        TestRequest,
+        TestResponse
+      >();
+
+      await endpointAdapter(request);
+
+      expect(mockedHttpClient).toHaveBeenCalledTimes(1);
+      expect(mockedHttpClient).toHaveBeenCalledWith(options.endpoint, {
+        parameters: request
+      });
+    });
+
+    // eslint-disable-next-line max-len
+    it('should combine the provided EndpointAdapterOptions.endpoint with the requestOptions.endpoint when the former is a function', async () => {
+      const requestOptions: RequestOptions = {
+        endpoint: 'https://api.empathy.co/test'
+      };
+      const endpoint: Mapper<string, string> = from => from.replace(/test/g, 'mapped');
+      const expectedEndpoint = endpoint(requestOptions.endpoint!, {});
```

Comment:
I have seen this multiple times already. Why you define expected values at the top of the test? That is confusing. Usually we structure tests following the AAA pattern (I thought this pattern was called differently honestly, I had to check it):

- Arrange: Do the necessary preparations
- Act: Execute the code that we are testing
- Assert: Run the needed assertions over the result of the executed code

By defining expected values at the top of the test is like you are mixing intentions.

Reply:
You're totally right. Thanks for the sharing the pattern!